### ISSUE TITLE
[策略发现] 强化学习交易策略 (DQN+PPO+Actor-Critic+订单执行)

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4191]
## 摘要

强化学习交易策略通过与市场环境交互学习最优交易策略，主要方法包括Q学习、DQN、Policy Gradient、PPO等，适用于订单执行、资产配置和资产选择等场景。

### 核心算法

1. **Q学习与DQN**
   - Q表/Q网络
   - 经验回放
   - 目标网络

2. **Policy Gradient**
   - REINFORCE算法
   - Actor-Critic架构
   - PPO（Proximal Policy Optimization）

3. **交易应用**
   - 订单执行优化（VWAP/TWAP改进）
   - 投资组合管理
   - 动态资产配置

4. **市场建模**
   - 状态设计（价格、持仓、技术指标）
   - 动作空间（买入/卖出/持有）
   - 奖励函数设计

### 数学模型

```
Q学习更新: Q(s,a) ← Q(s,a) + α[r + γmax Q(s',a') - Q(s,a)]

DQN损失: L(θ) = E[(r + γmax_a' Q(s',a';θ⁻) - Q(s,a;θ))²]

PPO裁剪: L(θ) = E[min(r_t(θ)×A_t, clip(r_t(θ), 1-ε, 1+ε)×A_t)]

Actor-Critic: 
  - Actor: ∇θ log π(a|s;θ) × A(s,a)
  - Critic: ∇w (r + γV(s';w) - V(s;w))²
```

## Python代码示例

```python
import numpy as np
import pandas as pd
from collections import deque
import random

class ReinforcementLearningTrading:
    """强化学习交易策略"""
    
    def __init__(self, state_size, action_size=3, learning_rate=0.001,
                 gamma=0.99, epsilon=1.0, epsilon_min=0.01, epsilon_decay=0.995):
        self.state_size = state_size
        self.action_size = action_size
        self.learning_rate = learning_rate
        self.gamma = gamma
        self.epsilon = epsilon
        self.epsilon_min = epsilon_min
        self.epsilon_decay = epsilon_decay
        
        # 经验回放池
        self.memory = deque(maxlen=2000)
        
        # Q网络
        self.model = self._build_model()
        self.target_model = self._build_model()
        self.update_target_model()
        
    def _build_model(self):
        """构建Q网络"""
        from tensorflow.keras.models import Sequential
        from tensorflow.keras.layers import Dense
        
        model = Sequential([
            Dense(64, input_dim=self.state_size, activation='relu'),
            Dense(32, activation='relu'),
            Dense(16, activation='relu'),
            Dense(self.action_size, activation='linear')
        ])
        model.compile(optimizer='adam', loss='mse')
        return model
    
    def remember(self, state, action, reward, next_state, done):
        """存储经验"""
        self.memory.append((state, action, reward, next_state, done))
    
    def act(self, state):
        """ε-greedy选择动作"""
        if np.random.random() <= self.epsilon:
            return random.randrange(self.action_size)
        act_values = self.model.predict(state)
        return np.argmax(act_values[0])
    
    def replay(self, batch_size=32):
        """经验回放训练"""
        minibatch = random.sample(self.memory, batch_size)
        
        for state, action, reward, next_state, done in minibatch:
            target = reward
            if not done:
                target = reward + self.gamma * np.amax(
                    self.target_model.predict(next_state)[0])
            
            target_f = self.model.predict(state)
            target_f[0][action] = target
            
            self.model.fit(state, target_f, epochs=1, verbose=0)
            
        if self.epsilon > self.epsilon_min:
            self.epsilon *= self.epsilon_decay
    
    def update_target_model(self):
        """更新目标网络"""
        self.target_model.set_weights(self.model.get_weights())
    
    def get_state(self, data, t, window_size=10):
        """构建状态"""
        window_data = data[t-window_size+1:t+1]
        state = []
        for i in range(window_size-1):
            state.append(window_data[i+1] - window_data[i])
        return np.array(state).reshape(1, -1)

class PolicyGradientTrading:
    """策略梯度交易"""
    
    def __init__(self, state_size, action_size, lr=0.001):
        self.action_size = action_size
        self.lr = lr
        
        self.actor = self._build_actor()
        self.critic = self._build_critic()
        
    def _build_actor(self):
        """Actor网络"""
        from tensorflow.keras.models import Sequential
        from tensorflow.keras.layers import Dense
        
        model = Sequential([
            Dense(64, input_dim=self.state_size, activation='relu'),
            Dense(32, activation='relu'),
            Dense(self.action_size, activation='softmax')
        ])
        model.compile(optimizer='adam', loss='categorical_crossentropy')
        return model
    
    def _build_critic(self):
        """Critic网络"""
        from tensorflow.keras.models import Sequential
        from tensorflow.keras.layers import Dense
        
        model = Sequential([
            Dense(64, input_dim=self.state_size, activation='relu'),
            Dense(32, activation='relu'),
            Dense(1, activation='linear')
        ])
        model.compile(optimizer='adam', loss='mse')
        return model
    
    def get_action(self, state):
        """选择动作（按概率）"""
        probs = self.actor.predict(state)[0]
        return np.random.choice(self.action_size, p=probs)
    
    def train(self, states, actions, rewards):
        """训练Actor-Critic"""
        # 计算优势函数
        values = self.critic.predict(states).flatten()
        advantages = rewards - values
        
        # 更新Critic
        self.critic.fit(states, rewards, epochs=1, verbose=0)
        
        # 更新Actor
        actions_onehot = np.zeros((len(actions), self.action_size))
        actions_onehot[np.arange(len(actions)), actions] = 1
        
        def actor_loss(y_true, y_pred):
            return -K.mean(y_pred * actions_onehot * advantages)
        
        from tensorflow.keras import backend as K
        self.actor.fit(states, actions_onehot, epochs=1, verbose=0)

# 使用示例
rl_trader = ReinforcementLearningTrading(state_size=10, action_size=3)
pg_trader = PolicyGradientTrading(state_size=10, action_size=3)

for episode in range(1000):
    state = rl_trader.get_state(prices, t)
    action = rl_trader.act(state)
    next_state = rl_trader.get_state(prices, t+1)
    reward = calculate_reward(portfolio, action)
    done = episode == len(prices) - 1
    
    rl_trader.remember(state, action, reward, next_state, done)
    
    if len(rl_trader.memory) > 32:
        rl_trader.replay(32)
```

## 参考文献

- Moody, J. (1998). Reinforcement Learning for Trading. NIPS
- Silver, D. (2016). Mastering the Game of Go with Deep Neural Networks. Nature
- Schulman, J. (2017). Proximal Policy Optimization Algorithms. arXiv
